service_management/app/repositories/analysis/dao/AnalysisDao.scala

```diff
@@ -190,21 +241,51 @@ class AnalysisDao @Inject()(
   }
 
   /**
-   * Adds a new result to the analysis with the given EventId.
+   * Usefull method for locating the result for a specific analysis event.
    *
-   * @param id The EventId of the analysis that has a new result.
-   * @param res The AnalysisResult to add.
-   * @return eventually a result with the database ID of the saved result.
+   * @param id the EventId of the analysis event to fetch the result for
+   * @return a result that may or may not contain the AnalysisResult
    */
-  def insertResult(id: EventId, res: AnalysisResult): Future[MusitResult[Long]] = {
-    val action = insertResultAction(asResultTuple(id, res))
+  def findResultFor(id: EventId): Future[MusitResult[Option[AnalysisResult]]] = {
+    val q = resultTable.filter(_.eventId === id).result.headOption.map(fromResultRow)
 
-    db.run(action).map(MusitSuccess.apply).recover {
+    db.run(q).map(MusitSuccess.apply).recover {
       case NonFatal(ex) =>
-        val msg = s"An unexpected error occurred inserting a result to analysis $id"
+        val msg = s"An unexpected error occurred looking up result for analysis $id"
         logger.error(msg, ex)
         MusitDbError(msg, Option(ex))
     }
   }
 
+  /**
+   * Adds or updates a result for the analysis with the given EventId.
+   *
+   * @param id  The EventId of the analysis that has a new result.
+   * @param res The AnalysisResult to add.
+   * @return eventually a result with the EventId the saved result belongs to.
+   */
+  def upsertResult(id: EventId, res: AnalysisResult): Future[MusitResult[EventId]] = {
+    val action = upsertResultAction(id, res)
+
+    db.run(action.transactionally)
+      .flatMap { numUpdated =>
+        val q = resultTable.filter(_.eventId === id).map(_.eventId)
+        db.run(q.result.headOption).map {
```

Comment:
Because it's an "upsert", the only time the "update" will fail is if there's a DB exception (which will cause a Future.failed). If the result doesn't exist, it will be added, if it exists it will be updated.
With only 2 levels deep blocks, for-comprehensions vs flatMap map... doesn't really matter all that much. Depends on the granularity of control you want.